lib/srv/git/forward.go

```diff
@@ -0,0 +1,590 @@
+	}
+	if c.AuthClient == nil {
+		return trace.BadParameter("missing parameter AuthClient")
+	}
+	if c.AccessPoint == nil {
+		return trace.BadParameter("missing parameter AccessPoint")
+	}
+	if c.Emitter == nil {
+		return trace.BadParameter("missing parameter Emitter")
+	}
+	if c.HostCertificate == nil {
+		return trace.BadParameter("missing parameter HostCertificate")
+	}
+	if c.ParentContext == nil {
+		return trace.BadParameter("missing parameter ParentContext")
+	}
+	if c.LockWatcher == nil {
+		return trace.BadParameter("missing parameter LockWatcher")
+	}
+	if c.SrcAddr == nil {
+		return trace.BadParameter("source address required to identify client")
+	}
+	if c.DstAddr == nil {
+		return trace.BadParameter("source address required to identify client")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// ForwardServer is an in-memory SSH server that forwards git commands to remote
+// git-hosting services like "github.com".
+type ForwardServer struct {
+	events.StreamEmitter
+	cfg    *ForwardServerConfig
+	logger *slog.Logger
+	auth   *srv.AuthHandlers
+	reply  *sshutils.Reply
+	id     string
+
+	// serverConn is the server side of the pipe to the client connection.
+	serverConn net.Conn
+	// clientConn is the client side of the pipe to the client connection.
+	clientConn net.Conn
+	// remoteClient is the client connected to the git-hosting service.
+	remoteClient *tracessh.Client
+
+	// waitExec receives exec response.
+	waitExec chan error
+
+	// verifyRemoteHost is a callback to verify remote host like "github.com".
+	// Can be overridden for tests. Defaults to verifyRemoteHost.
+	verifyRemoteHost ssh.HostKeyCallback
+	// makeRemoteSigner generates the client certificate for connecting to the
+	// remote server. Can be overridden for tests. Defaults to makeRemoteSigner.
+	makeRemoteSigner func(context.Context, *ForwardServerConfig, srv.IdentityContext) (ssh.Signer, error)
+}
+
+// Dial returns the client connection of the pipe
+func (s *ForwardServer) Dial() (net.Conn, error) {
+	return s.clientConn, nil
+}
+
+// NewForwardServer creates a new in-memory SSH server that forwards git
+// commands to remote git-hosting services like "github.com".
+func NewForwardServer(cfg *ForwardServerConfig) (*ForwardServer, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	serverConn, clientConn, err := utils.DualPipeNetConn(cfg.SrcAddr, cfg.DstAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	logger := slog.With(teleport.ComponentKey, teleport.ComponentForwardingGit,
+		"src_addr", cfg.SrcAddr.String(),
+		"dst_addr", cfg.DstAddr.String(),
+	)
+	s := &ForwardServer{
+		StreamEmitter:    cfg.Emitter,
+		cfg:              cfg,
+		serverConn:       serverConn,
+		clientConn:       clientConn,
+		logger:           logger,
+		reply:            sshutils.NewReply(logger),
+		id:               uuid.NewString(),
+		waitExec:         make(chan error, 1),
```

Comment:
This is shared between all channels on the connection, it's a logical race condition.